src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -0,0 +1,674 @@
+        /// before the place where we execute arbitrary code.
+        /// (i.e. do the reborrows)
+        fn before_arbitrary_code(&self) -> Result<LocStateProtPair, ()> {
+            let mut state = LocStateProtPair {
+                x: LocStateProt { state: self.x_initial_state, prot: true },
+                y: LocStateProt { state: self.y_initial_state, prot: self.y_protected },
+            };
+
+            if self.x_retag_first {
+                // y does not exist initially
+                state.x = state.x.perform_access(AccessKind::Read, AccessRelatedness::This)?;
+
+                if state.y.state.initialized {
+                    state = state.perform_test_access(
+                        &TestAccess { ptr: PtrSelector::Y, kind: AccessKind::Read },
+                        self.xy_rel,
+                    )?;
+                }
+            } else {
+                // x does not exist initially
+                if state.y.state.initialized {
+                    // this is actually a noop
+                    state.y = state.y.perform_access(AccessKind::Read, AccessRelatedness::This)?;
+                }
+
+                state = state.perform_test_access(
+                    &TestAccess { ptr: PtrSelector::X, kind: AccessKind::Read },
+                    self.xy_rel,
+                )?;
+            }
+            Ok(state)
+        }
+
+        /// Given the state immediately after the arbitrary code, execute
+        /// the pattern until the point where we want to do a distinguishability check
+        /// (i.e. maybe execute the spurious read)
+        fn after_arbitrary_code(
+            &self,
+            state: &LocStateProtPair,
+            spurious_read: bool,
+        ) -> Result<LocStateProtPair, ()> {
+            // `x` must still be protected by the time we perform the spurious read.
+            let state = if spurious_read {
+                state.perform_test_access(
+                    &TestAccess { ptr: PtrSelector::X, kind: AccessKind::Read },
+                    self.xy_rel,
+                )?
+            } else {
+                state.clone()
+            };
+            Ok(state)
+        }
+    }
+
+    #[test]
+    #[should_panic]
+    /// For each of the patterns described above, execute it once
+    /// as-is, and once with a spurious read inserted. Report any UB
+    /// in the target but not in the source.
+    fn test_all_patterns() {
+        let mut ok = 0;
+        let mut err = 0;
+        for pat in Pattern::exhaustive() {
+            precondition!(pat.xy_rel.allows_x_retag_first() || !pat.x_retag_first);
+            let Ok(common_0) = pat.before_arbitrary_code() else {
+                continue;
+            };
+            // `x` must stay protected, but the function of `y` might return here
+            for (common_1, opaque) in
+                common_0.applicable_opaque_code_without_ub::<!, ()>(pat.xy_rel)
```

Comment:
So... is `RetX` always `!` and `RetY` always `()`? Then why make all that code generic? Also further down the look I see `::<(), ()>`, why does that use a different `RetX`? Definitely all of these need comments explaining wht we are picking the types we are picking. This is very implicitly making some super important choices!